File: fortran/pr033.f90
! Problem 33
!
! 20 December 2002
!
! The fraction ^49/[98] is a curious fraction, as an inexperienced
! mathematician in attempting to simplify it may incorrectly believe
! that ^49/[98] = ^4/[8], which is correct, is obtained by cancelling
! the 9s.
!
! We shall consider fractions like, ^30/[50] = ^3/[5], to be trivial
! examples.
!
! There are exactly four non-trivial examples of this type of fraction,
! less than one in value, and containing two digits in the numerator
! and denominator.
!
! If the product of these four fractions is given in its lowest common
! terms, find the value of the denominator.
! 100

subroutine pr033

  use primes, only: gcd
  implicit none
  integer :: a, b, prod_n, prod_m, g

  prod_n = 1
  prod_m = 1

  do a = 10, 98
    do b = a+1, 99
      if (is_fraction(a, b)) then
        ! print *, a, b
        ! Multiple and reduce.

        prod_n = prod_n * a
        prod_m = prod_m * b

        g = gcd(prod_n, prod_m)
        if (g > 1) then
          prod_n = prod_n / g
          prod_m = prod_m / g
        end if
      end if
    end do
  end do

  print *, prod_m

contains

  !------------------------------------------------------------
  ! Return .true. if a/b forms one of the special fractions.
  function is_fraction(a, b)
    implicit none
    integer, intent(in) :: a, b
    logical :: is_fraction
    integer :: an, am, bn, bm

    an = a / 10
    am = mod(a, 10)
    bn = b / 10
    bm = mod(b, 10)

    is_fraction = ((an == bm .and. bn > 0 .and. am*b == bn*a) .or. &
      (am == bn .and. bm > 0 .and. an*b == bm*a))
  end function is_fraction

end subroutine pr033

File: fsharp/pr033.fs
// Problem 33
//
// 20 December 2002
//
//
// The fraction ^49/[98] is a curious fraction, as an inexperienced
// mathematician in attempting to simplify it may incorrectly believe that ^
// 49/[98] = ^4/[8], which is correct, is obtained by cancelling the 9s.
//
// We shall consider fractions like, ^30/[50] = ^3/[5], to be trivial
// examples.
//
// There are exactly four non-trivial examples of this type of fraction, less
// than one in value, and containing two digits in the numerator and
// denominator.
//
// If the product of these four fractions is given in its lowest common
// terms, find the value of the denominator.

let isFrac a b =
    let an = a / 10
    let am = a % 10
    let bn = b / 10
    let bm = b % 10
    (an = bm && bn > 0 && am*b = bn*a) ||
        (am = bn && bm > 0 && an*b = bm*a)

let rec gcd a b =
    if b = 0 then a
    else gcd b (a % b)

// Rational multiplation that reduces the fraction
let ratMult (a, b) (c, d) =
    let n = a * c
    let m = b * d
    let common = gcd m n
    (n / common, m / common)

let pr33 () =
    let (_, m) =
        seq {
            for a = 10 to 99 do
                for b = a+1 to 99 do
                    if isFrac a b then yield (a, b) }
        |> Seq.fold ratMult (1, 1)
    m

printfn "%A" (pr33 ())
